linfa-linear/src/elastic.rs

```diff
@@ -0,0 +1,572 @@
+        self
+    }
+
+    /// Set the tolerance which is the minimum absolute change in any of the
+    /// model parameters needed for the parameter optimization to continue.
+    ///
+    /// Defaults to `1e-4` if not set
+    pub fn tolerance(mut self, tolerance: F) -> Self {
+        self.elastic_net = self.elastic_net.tolerance(tolerance);
+        self
+    }
+
+    /// Set the maximum number of iterations for the optimization routine.
+    ///
+    /// Defaults to `1000` if not set
+    pub fn max_iterations(mut self, max_iterations: u32) -> Self {
+        self.elastic_net = self.elastic_net.max_iterations(max_iterations);
+        self
+    }
+
+    /// Fit a ridge regression model given a feature matrix `x` and a target
+    /// variable `y`.
+    ///
+    /// The feature matrix `x` must have shape `(n_samples, n_features)`
+    ///
+    /// The target variable `y` must have shape `(n_samples)`
+    ///
+    /// Returns a `FittedElasticNet` object which contains the fitted
+    /// parameters and can be used to `predict` values of the target variable
+    /// for new feature values.
+    pub fn fit(&self, x: &Array2<F>, y: &Array1<F>) -> Result<FittedElasticNet<F>, String> {
+        self.elastic_net.fit(x, y)
+    }
+}
+
+/// View the fitted parameters and make predictions with a fitted
+/// elastic net model
+impl<F: Float + FromPrimitive + ScalarOperand> FittedElasticNet<F> {
+    /// Given an input matrix `X`, with shape `(n_samples, n_features)`,
+    /// `predict` returns the target variable according to elastic net
+    /// learned from the training data distribution.
+    pub fn predict(&self, x: &Array2<F>) -> Array1<F> {
+        x.dot(&self.parameters) + self.intercept
+    }
+
+    /// Get the fitted parameters
+    pub fn parameters(&self) -> &Array1<F> {
+        &self.parameters
+    }
+
+    /// Get the fitted intercept, 0. if no intercept was fitted
+    pub fn intercept(&self) -> F {
+        self.intercept
+    }
+
+    /// Get the number of steps taken in optimization algorithm
+    pub fn n_steps(&self) -> u32 {
+        self.n_steps
+    }
+
+    /// Get the duality gap at the end of the optimization algorithm
+    pub fn duality_gap(&self) -> F {
+        self.duality_gap
+    }
+}
+
+fn coordinate_descent<F: AbsDiffEq + Float + FromPrimitive + ScalarOperand + NumAssignOps>(
+    x: &Array2<F>,
+    y: &Array1<F>,
+    tol: F,
+    max_steps: u32,
+    l1_ratio: F,
+    penalty: F,
+) -> (Array1<F>, F, u32) {
+    let n_samples = F::from(x.shape()[0]).unwrap();
+    let n_features = x.shape()[1];
+    // the parameters of the model
+    let mut w = Array1::<F>::zeros(n_features);
+    // the residuals: `y - X*w` (since w=0, this is just `y` for now),
+    // the residuals are updated during the algorithm as the parameters change
+    let mut r = y.clone();
+    let mut n_steps = 0u32;
+    let norm_cols_x = x.map_axis(Axis(0), |col| {
+        col.fold(F::zero(), |sum_sq, &x| sum_sq + x * x)
```

Comment:
```rust
col.mapv_into(|x| x*x).sum()
```
should also work here

Reply:
I think this requires ownership/mutability (for `mapv_inplace`) of the underlying data, but I'm working on a shared reference here. I would also not want to copy the column previous to manipulating it to avoid the extra allocation.

Reply:
But I guess `col.dot(&col)` should work.

Reply:
yeah that's true, you could do `x.mapv(|y| y*y).sum_axis(Axis(1))`, but this would copy the `ndarray` as well